### PR TITLE
tplgtool2.py: show PCM names on graph output

### PR DESCRIPTION
**Pull request by marc-hb**

### CODE REVIEW

**Comment by marc-hb** [comment]
Seems to work great... for topology1. There's no "PCMxP" in topology2.

topo1 with PCM name

**Comment by marc-hb** [comment]
Besides `host.copier.7.capture` having no corresponding PCM in https://sof-ci.01.org/softestpr/PR1054/build420/devicetest/index.html?model=TGLU_RVP_NOCODEC_IPC4ZPH&testcase=verify-sof-firmware-load and in some MTL devices, the graphs looks much better with PCM names.

In any case it seems a bit weird to have to parse the widget name... is there no simpler widget->pcm connection in the topology?

**Review by RanderWang — DISMISSED** [review]
LGTM

File: tools/tplgtool2.py [hunk 1]
```diff
@@ -864,6 +864,34 @@ def node_sublabel(self, widget: Container) -> str:
...
+
+        pcm_id = int(match["num"])
+        pcm_names = [ pcm.pcm_name for pcm in self._tplg.pcm_list if pcm.pcm_id == pcm_id ]
+
+        n_l = len(pcm_names)
+        # A duplicate pcm_id is a very severe issue
+        assert n_l <= 1, f"Unexpected number of PCMs matching pcm_id={pcm_id}: {n_l} ({pcm_names})"
+
+        return pcm_names[0] if n_l == 1 else f"No PCM with ID={pcm_id}"
+
```
Comment (plbossart) [inline tools/tplgtool2.py:894]: IIRC @ranj063 already did some changes in the topology2 naming, not sure if these changes are based on top of the topology2 changes?

In addition, while the name of the PCM device is nice, the PCM device number is the most useful for testing. How difficult would it be to add the hw: ID?

**Review by ranj063 — DISMISSED** [review]

File: tools/tplgtool2.py (hunk 1 above)
Comment (ranj063) [reply tools/tplgtool2.py:894]: > IIRC @ranj063 already did some changes in the topology2 naming, not sure if these changes are based on top of the topology2 changes?
> 
> In addition, while the name of the PCM device is nice, the PCM device number is the most useful for testing. How difficult would it be to add the hw: ID?

@plbossart I think this is a good change on top of what I did with the naming. With a name like host-copier.1.playback you've already got the PCM ID i.e. 1 and this adds the name of the PCM next to it

File: tools/tplgtool2.py [hunk 2]
```diff
@@ -864,6 +864,34 @@ def node_sublabel(self, widget: Container) -> str:
...
+        wname = widget.name
+
+        match = self.HOST_COPIER_NUM_REGEX.match(wname)
+        if match is None:
+            match = self.PCM_NUM_REGEX.match(wname)
+
+        if match is None:
+            # Don't crash the entire program when passed a wrong input;
+            # put an error message in the graph instead.
+            return f'No PCM or host-copier number found in "{wname}"'
```
Comment (marc-hb) [inline tools/tplgtool2.py:884]: From @ranj063 : this is a topology error, make it a hard fail.

**Review by plbossart — APPROVED** [review]

**Comment by marc-hb** [comment]
I started running the new version manually on all .tplg files and I'm finding new errors:

```
topology/topology1/production/sof-tgl-nocodec.tplg

pcm=Port0;id=0;type=duplex;fmt=S16_LE;rate_min=48000;rate_max=48000;ch_min=2;ch_max=2;
pcm=Port1;id=1;type=duplex;fmt=S16_LE;rate_min=48000;rate_max=48000;ch_min=2;ch_max=2;
pcm=Port2;id=2;type=duplex;fmt=S16_LE;rate_min=48000;rate_max=48000;ch_min=2;ch_max=2;
pcm=DMIC;id=10;type=capture;fmt=S16_LE;rate_min=48000;rate_max=48000;ch_min=4;ch_max=4;
pcm=DMIC16kHz;id=11;type=capture;fmt=S16_LE;rate_min=16000;rate_max=16000;ch_min=4;ch_max=4;


AssertionError: Zero PCM with ID=3 found for widget=PCM3P

```

**Comment by marc-hb** [comment]
```
topology/topology2/sof-ace-tplg/sof-mtl-max98357a-rt5682.tplg

pcm=EchoRef;id=27;type=capture;fmt=S16_LE;rate_min=48000;rate_max=48000;ch_min=2;ch_max=2;
pcm=DMIC0 RTC AEC;id=27;type=capture;fmt=S32_LE;rate_min=48000;rate_max=48000;ch_min=4;ch_max=4;
pcm=DMIC0 RTC AEC NS;id=28;type=capture;fmt=S32_LE;rate_min=48000;rate_max=48000;ch_min=4;ch_max=4;


AssertionError: More than one PCM has pcm_id=27: 2 (['EchoRef', 'DMIC0 RTC AEC'])

```

**Review by RanderWang — DISMISSED** [review]
LGTM

**Comment by marc-hb** [comment]
graph still missing in https://sof-ci.01.org/softestpr/PR1054/build568/devicetest/index.html?model=APL_UP2_NOCODEC&testcase=verify-tplg-binary, there's probably a `set -e` in CI somewhere.

**Comment by marc-hb** [comment]
SOFCI TEST

**Comment by marc-hb** [comment]
SOFCI TEST

**Comment by marc-hb** [comment]
SOFCI TEST

**Comment by marc-hb** [comment]
For a while I naively thought I could quickly fix the problematic topologies above - and then merge this. I have neither enough time nor knowledge for that. So the most recent and... perfect-at-last revision does a "soft fail": it still produces the graph even when a PCM id is not found. That's what it used to do before this PR and graph is probably useful for troubleshooting anyway so I think it's the best design after all. An error message is printed on stdout and placed in the graph too, there's also a non-zero exit status.

**Review by RanderWang — APPROVED** [review]
LGTM

**Comment by marc-hb** [comment]
thanks @RanderWang for your patience through the error-handling back and forth!

**Pull request merged**